cpp/include/cudf/join.hpp

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace cudf {
+// joins
+
+/** 
+ * @brief  Performs an inner join on the specified columns of two
+ * tables (left, right)
+ *
+ * @example TableA a: {0, 1, 2}
+ *          TableB b: {1, 2, 3}, a: {1, 2, 5}
+ *          left_on: {0}
+ *          right_on: {1}
+ *          columns_in_common: { {0, 1} }
```

Comment:
I'm confused by columns_in_common here. The doc below says this is
> a vector of pairs of column indices into `left_on` and `right_on`
If that were the case, wouldn't columns_in_common need to be {0, 0}? In the example, it seems that they are actually indices into `left` and `right`

Reply:
Better rewording could be 

> a vector of pairs of column indices from `left_on` and `right_on`

For eg. in this case, the pairs in columns_in_common cannot be anything but `{0,1}` or `{}`

Reply:
@trevorsm7  

Pairs of indices depends on which column you are taking as key columns, from python you can choose columns comparable type to make a join. In this example, the chosen columns are column_0 from TableA and column_1 from TableB. 

And normally `columns_in_common` deals with columns with same name which is linked from python side, so if the column name of the joining columns are same, then their pair indices information will be in `columns_in_common`. And in python column can be at any position in a table.

Reply:
> I'm confused by columns_in_common here. The doc below says this is
> 
> > a vector of pairs of column indices into `left_on` and `right_on`
> 
> If that were the case, wouldn't columns_in_common need to be {0, 0}? In the example, it seems that they are actually indices into `left` and `right`

I think I got your point, `columns_in_common` have the same indices stored in `left_on` and `right_on` and not index of/ index into to `left_on` and `right_on`.